hedy.py

```diff
@@ -2471,7 +2476,9 @@ def get_remaining_rules(orig_def, sub_def):
     return result_cmd_list
 
 
-def create_grammar(level, lang="en"):
+# this is only a couple of MB in total, safe to cache
```

Comment:
> Is it? 40 languages * 18 levels * 5 processes per instance * 1MB = ~3.5GB, out of the 1GB the instance has.

These are the plain text lark strings that take about 10KB each, so +/- 30MB, not the parser objects which are 1MB each.

> Also, is there value in caching it if we cache the parser already?

I profiled the get_parser function yesterday and if I recall correctly it takes about 50ms to run the create_grammar function, which is not a lot, but it can be <1ms with this cache.

The grammar is an input for the cache key (to access the disk cache), so if we remove the `@lru_cache` below, or if we have a cache miss there, we have to run the 50ms `create_grammar` function. That's why adding it is a good idea IMHO.

Reply:
> These are the plain text lark strings that take about 10KB each

Oh I see. Gotcha!

> I profiled the get_parser function yesterday and if I recall correctly it takes about 50ms to run the create_grammar function, which is not a lot, but it can be <1ms with this cache.

I understand. But if we only ever call `create_grammar` once for each parser (because the parser itself ends up pickled anyway), we're not saving the 50ms but we *are* paying the memory cost.

Reply:
As a safety precaution, we use the grammar -> to create a hash -> to store/retrieve the pickled parser to disk for exactly that grammar. Reasoning: if we change anything in the grammar, we should create a new parser. This is what Lark also does under the hood. Now we could drive without seat belts here, because in production we know the grammar files won't change for the language/level combination, but I feel like that could lead to hard-to-debug issues in the future.

Reply:
I see. The only hard-to-debug issue that I can think of is, on a developer workstation, changing the grammar files without restarting the server will sometimes load the new version and sometimes the old. I think that's a manageable risk.

Do you seen any others?

Reply:
Can't we cache the grammar hash instead of the grammar itself? Much smaller, same effect.

Reply:
I like this idea but it's pretty difficult to implement in a readable way. I would prefer this to be done in a follow-up issue, seeing as we are already reducing memory usage by 100x in this PR.